src/parser.cpp

```diff
@@ -1,30 +1,267 @@
-#pragma once
-
 #include "ada.h"
-#include "unicode.cpp"
-#include "scheme.cpp"
 
-#include <cctype>
+#include "checkers.cpp"
+
 #include <algorithm>
+#include <cctype>
+#include <cmath>
+#include <cstdlib>
+#include <string>
+
+namespace ada::parser {
+
+  /**
+   * @see https://url.spec.whatwg.org/#concept-domain-to-ascii
+   */
+  parser_result<std::string_view> domain_to_ascii(const std::string_view input, bool be_strict) {
+    // TODO: Implement this
+    return std::make_tuple(std::nullopt, false);
+  }
+
+  /**
+   * @see https://url.spec.whatwg.org/#concept-opaque-host-parser
+   */
+  parser_result<std::string_view> parse_opaque_host(std::string_view input) {
+    bool has_validation_error = false;
+
+    for (auto i = input.begin(); i < input.end(); i++) {
+      // If input contains a forbidden host code point, validation error, return failure.
+      if (ada::unicode::is_in_code_points(*i, ada::unicode::FORBIDDEN_HOST_CODE_POINTS)) {
+        return std::make_tuple(std::nullopt, true);
+      }
+
+      // Optimization: No need to continue the loop if we have a validation error
+      if (has_validation_error) {
+        break;
+      }
+
+      // If input contains a code point that is not a URL code point and not U+0025 (%), validation error.
+      if (!ada::unicode::is_ascii_alphanumeric(*i) && *i != '%') {
+        has_validation_error = true;
+      }
+
+      // If input contains a U+0025 (%) and the two code points following it are not ASCII hex digits, validation error.
+      if (*i == '%' && std::distance(i, input.end()) < 2 && (!ada::unicode::is_ascii_hex_digit(i[1]) || !ada::unicode::is_ascii_hex_digit(i[2]))) {
+        has_validation_error = true;
+      }
+    }
+
+    // Return the result of running UTF-8 percent-encode on input using the C0 control percent-encode set.
+    std::string_view result = ada::unicode::utf8_percent_encode(input, ada::character_sets::C0_CONTROL_PERCENT_ENCODE);
+
+    return std::make_tuple(result, has_validation_error);
+  }
+
+  /**
+   * @see https://url.spec.whatwg.org/#concept-ipv4-parser
+   */
+  parser_result<std::string_view> parse_ipv4(std::string_view input) {
```

Comment:
This is another function which we might be able to make much more efficient. Let us revisit later.

Reply:
I agree. Eventually, we need multiple benchmarks testing ipv4, ipv6, file paths and regular URLs.